src/io/rsm/raft.hpp

```diff
@@ -182,7 +181,7 @@ struct PendingClientRequest {
 
 struct Leader {
   std::map<Address, FollowerTracker> followers;
-  std::unordered_map<LogIndex, PendingClientRequest> pending_client_requests;
+  std::map<LogIndex, PendingClientRequest> pending_client_requests;
```

Comment:
with the current sketch code it is not necessary, but soon we will iterate over the `pending_client_requests` to handle timeouts, and at that point, it will need to be a `std::map` for determinism. 

The benefits of a hashmap over a tree mostly start becoming significant when the tree structure involves more random memory for nodes (compared to the single allocation of the hashmap) than will fit in L1 cache, and then even more different at the L2 cache boundary. It's possible that we will eventually want to conditionally compile this to use a hashmap for production and a tree for simulation, but in general I estimate that the measurable difference is not going to impact us much in the near term.

Reply:
In the C++ community (maybe because of implementation details, I am not sure) a very well known rule of thumb is prefer unordered containers (sets and maps) over ordered ones unless you want to capitalize on the ordered or "node-based" behavior of the container. Therefore on the hot-path (this is probably on the hot-path), I would always prefer the unordered containers if possible.

I think the determinism doesn't depend on the ordered behavior of the map, only on the deterministic behavior, which is true for both `unordered_map` and `map`, or at least I don't see how the `unordered_map` here would affect the deterministic behavior of the simulation. As we totally control the communication between threads (by wait for all threads to block) and delivering the messages one-by-one, I think the order of how the messages are handled on a single thread shouldn't affect the deterministic behavior of the simulation.

Could you please explain why an `unordered_map` would hurt the simulation once we start to iterate over the pending requests?

Reply:
yeah, [the performance advantages of hashmaps are clear](https://twitter.com/sadisticsystems/status/1594299358303977472), but many industrial hashmap implementations incorporate some amount of runtime entropy to prevent collision attacks. I don't know about the specifics of our specific version of libstd++, but it's standard practice in the systems engineering community to avoid hashmaps in deterministic contexts.

Reply:
it should also be noted that when iterating in lexicographic order over tree, the point operation performance is often better than a hashmap, because you're likely to already have most of the nodes in the access path in L1 cache, so for this use case maps may be better anyway